sql/parser/eval.go

```diff
@@ -519,33 +520,26 @@ var evalTupleEQ = cmpOp{
 }
 
 var evalTupleIN = cmpOp{
-	fn: func(arg, values Datum, _ *interface{}) (DBool, error) {
+	fn: func(arg, values Datum, _ *interface{}) (in DBool, err error) {
 		if arg == DNull {
 			return DBool(false), nil
 		}
 
 		vtuple := values.(DTuple)
 
-		// TODO(pmattis): If we're evaluating the expression multiple times we should
-		// use a map when possible. This works as long as arg is not a tuple. Note
-		// that the usage of the map is currently disabled via the "&& false" because
-		// building the map is a pessimization if we're only evaluating the
-		// expression once. We need to determine when the expression will be
-		// evaluated multiple times before enabling. Also need to figure out a way to
-		// use the map approach for tuples. One idea is to encode the tuples into
-		// strings and then use a map of strings.
-		if _, ok := arg.(DTuple); !ok && false {
-			m := make(map[Datum]struct{}, len(vtuple))
-			for _, val := range vtuple {
-				if reflect.TypeOf(arg) != reflect.TypeOf(val) {
-					return DummyBool, fmt.Errorf("unsupported comparison operator: <%s> %s <%s>",
-						arg.Type(), EQ, val.Type())
+		if _, ok := arg.(DTuple); !ok {
+			defer func() {
+				if r := recover(); r != nil {
```

Comment:
you don't need to scan the entire tuple; you can check type sanity in the binary search comparator